src/test/rpc_wallet_tests.cpp

```diff
@@ -439,6 +478,544 @@ BOOST_AUTO_TEST_CASE(rpc_wallet_z_importexport)
+
+    // Add operation 2 and 3 to the queue
+    q->addOperation(op2);
+    std::shared_ptr<AsyncRPCOperation> op3(new MockSleepOperation(1000));
+    q->addOperation(op3);
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+    BOOST_CHECK_EQUAL(op2->isExecuting(), true);
+    op2->cancel();  // too late, already executing
+    op3->cancel();
+    std::this_thread::sleep_for(std::chrono::milliseconds(3000));
+    BOOST_CHECK_EQUAL(op2->isSuccess(), true);
+    BOOST_CHECK_EQUAL(op2->isCancelled(), false);
+    BOOST_CHECK_EQUAL(op3->isCancelled(), true);
+    
+   
+    v = q->getAllOperationIds();
+    std::copy( v.begin(), v.end(), std::inserter( opids, opids.end() ) );
+    BOOST_CHECK(opids.size() == 3);
+    BOOST_CHECK(opids.count(id1)==1);
+    BOOST_CHECK(opids.count(id2)==1);
+    BOOST_CHECK(opids.count(op3->getId())==1);
+    q->finishAndWait();
+}
+
+
+// The CountOperation will increment this global
+std::atomic<int64_t> gCounter(0);
+
+class CountOperation : public AsyncRPCOperation {
+public:
+    CountOperation() {}
+    virtual ~CountOperation() {}
+    virtual void main() {  
+        set_state(OperationStatus::EXECUTING);
+        gCounter++;
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+        set_state(OperationStatus::SUCCESS);
+    }
+};
+
+// This tests the queue waiting for multiple workers to finish
+BOOST_AUTO_TEST_CASE(rpc_wallet_async_operations_parallel_wait)
+{
+    gCounter = 0;
+    
+    std::shared_ptr<AsyncRPCQueue> q = std::make_shared<AsyncRPCQueue>();
+    q->addWorker();
+    q->addWorker();
+    q->addWorker();
+    q->addWorker();
+    BOOST_CHECK(q->getNumberOfWorkers() == 4);
+
+    int64_t numOperations = 10;     // 10 * 1000ms / 4 = 2.5 secs to finish
+    for (int i=0; i<numOperations; i++) {
+        std::shared_ptr<AsyncRPCOperation> op(new CountOperation());
+        q->addOperation(op);
+    }
+
+    std::vector<AsyncRPCOperationId> ids = q->getAllOperationIds();
+    BOOST_CHECK(ids.size()==numOperations);
+    q->finishAndWait();
+    BOOST_CHECK_EQUAL(q->isFinishing(), true);
+    BOOST_CHECK_EQUAL(numOperations, gCounter.load());
+}
+
+// This tests the queue shutting down immediately
+BOOST_AUTO_TEST_CASE(rpc_wallet_async_operations_parallel_cancel)
+{
+    gCounter = 0;
+    
+    std::shared_ptr<AsyncRPCQueue> q = std::make_shared<AsyncRPCQueue>();
+    q->addWorker();
+    q->addWorker();
+    BOOST_CHECK(q->getNumberOfWorkers() == 2);
+
+    int numOperations = 10000;  // 10000 seconds to complete
+    for (int i=0; i<numOperations; i++) {
+        std::shared_ptr<AsyncRPCOperation> op(new CountOperation());
+        q->addOperation(op);
+    }
+    std::vector<AsyncRPCOperationId> ids = q->getAllOperationIds();
+    BOOST_CHECK(ids.size()==numOperations);
+    q->closeAndWait();
+    
+    // the shared counter should equal the number of successful operations.
+    BOOST_CHECK_NE(numOperations, gCounter.load());
```

Comment:
This might have a race condition if some of the operations had not incremented the counter by this point. (A pause in the test thread before the call to ``q->closeAndWait()`` would suffice for now.)